lib/stdlib/src/gen_server.erl

```diff
@@ -2243,33 +2232,73 @@ init_it(Mod, Args) ->
 
 loop(ServerData, State, {continue, Continue} = Msg, Debug) ->
     Reply = try_handle_continue(ServerData, State, Continue),
+    From  = undefined,
     case Debug of
         [] ->
-            handle_common_reply(ServerData, State, Msg, undefined, Reply);
+            handle_common_reply(ServerData, State, Msg, From, Reply);
         _ ->
             Debug1 = sys:handle_debug(Debug, fun print_event/3, ServerData#server_data.name, Msg),
-            handle_common_reply(ServerData, State, Msg, undefined, Reply, Debug1)
+            handle_common_reply(ServerData, State, Msg, Debug1, From, Reply)
     end;
+%%
+loop(ServerData, State, LoopAction, Debug) ->
+    case LoopAction of
+        {timeout_zero, TimeoutMsg} ->
+            decode_msg(ServerData, State, TimeoutMsg, Debug, undefined, false);
+        {timeout, Timer, HibInf} ->
+            loop(ServerData, State, HibInf, Debug, Timer);
+        HibT ->
+            loop(ServerData, State, HibT, Debug, undefined)
+    end.
 
-loop(ServerData, State, hibernate, Debug) ->
-    proc_lib:hibernate(?MODULE,wake_hib,[ServerData, State, Debug]);
-
-loop(#server_data{hibernate_after=HibernateAfterTimeout} = ServerData, State, infinity, Debug) ->
+loop(ServerData, State, hibernate, Debug, Timer) ->
+    receive
+	Msg ->
+	    erlang:garbage_collect(),
+	    decode_msg(ServerData, State, Msg, Debug, Timer, true)
+    after 0 ->
+	loop_hibernate(ServerData, State, Debug, Timer)
+    end;
+%%
+loop(#server_data{hibernate_after = HibernateAfterTimeout} = ServerData, State, infinity, Debug, Timer) ->
     receive
-        Msg ->
-            decode_msg(ServerData, State, Msg, infinity, Debug, false)
+	Msg ->
+	    decode_msg(ServerData, State, Msg, Debug, Timer, false)
     after HibernateAfterTimeout ->
-            loop(ServerData, State, hibernate, Debug)
+	loop_hibernate(ServerData, State, Debug, Timer)
     end;
-
-loop(ServerData, State, Time, Debug) ->
+%%
+loop(ServerData, State, Time, Debug, Timer)
+  when ?is_rel_timeout(Time) ->
     Msg = receive
-              Input ->
-                  Input
-          after Time ->
-                  timeout
-          end,
-    decode_msg(ServerData, State, Msg, Time, Debug, false).
+	      Input ->
+		  Input
+	  after Time ->
+	      timeout
+	  end,
+    decode_msg(ServerData, State, Msg, Debug, Timer, false).
+
+loop_hibernate(ServerData, State, Debug, Timer) ->
+    erlang:hibernate(),
+    loop_wakeup(ServerData, State, Debug, Timer).
+
+loop_wakeup(ServerData, State, Debug, Timer) ->
```

Comment:
I don't see any point in `loop_wakeup/4` vs. inlining it in `loop_hibernate/4`.

Reply:
I like the new `hibernate/0` and how the code using it becomes simpler - it is a more convenient concept!

Reply:
For the `gen_server` code by itself, yes (and I had it this way at first). But it may hide a bug: in the tests, checking if a process is in hibernation is done by comparing the result of `erlang:process_info(Pid, current_function)` against `{gen_server, loop_hibernate, 4}`. This result is returned both when the process _is_ in hibernation (ie, in `erlang:hibernate/0`), but _also_ if it is _not_ hibernating any more but instead waiting in the following `receive` (which would be a bug).

I agree that this should not happen, in that if a process gets "past" the `erlang:hibernate/0`, it _must_ be because there is a message. Not sure we should cover this here though, because if that ever happens, the bug would not be in `gen_server`.

Reply:
Right, I see!  That might prove a useful debugging property one day.